styx-api-service/src/main/java/com/spotify/styx/api/ResourceResource.java

```diff
@@ -0,0 +1,171 @@
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Stream;
+import okio.ByteString;
+
+public final class ResourceResource {
+
+  public static final String BASE = "/resources";
+  public static final ObjectMapper OBJECT_MAPPER = Json.OBJECT_MAPPER;
+
+  private final Storage storage;
+
+  public ResourceResource(Storage storage) {
+    this.storage = Objects.requireNonNull(storage);
+  }
+
+  public Stream<? extends Route<? extends AsyncHandler<? extends Response<ByteString>>>> routes() {
+    final List<Route<AsyncHandler<Response<ByteString>>>> v0 = Arrays.asList(
+        Route.with(
+            json(), "GET", BASE,
+            rc -> getResources()),
+        Route.with(
+            json(), "POST", BASE,
+            rc -> postResource(rc.request())),
+        Route.with(
+            json(), "GET", BASE + "/<rid>",
+            rc -> getResource(arg("rid", rc))),
+        Route.with(
+            json(), "DELETE", BASE + "/<rid>",
+            rc -> deleteResource(arg("rid", rc))),
+        Route.with(
+            json(), "PATCH", BASE + "/<rid>",
+            rc -> patchResource(arg("rid", rc), rc.request()))
+    );
+
+    return cat(
+        v0.stream().map(r -> r.withPrefix(Api.Version.V0.prefix()))
+    );
+  }
+
+  private Response<List<Resource>> getResources() {
+    try {
+      return Response.forPayload(storage.getResources());
+    } catch (IOException e) {
+      return Response
+          .forStatus(
+              Status.INTERNAL_SERVER_ERROR.withReasonPhrase("Failed to get resources."));
+    }
+  }
+
+  private Response<Resource> getResource(String id) {
+    try {
+      return Response.forPayload(storage.getResource(id));
+    } catch (IOException e) {
+      return Response
+          .forStatus(
+              Status.INTERNAL_SERVER_ERROR.withReasonPhrase("Failed to get resource."));
+    }
+  }
+
+  private Response<Resource> deleteResource(String id) {
+    try {
+      storage.deleteResource(id);
+    } catch (ResourceNotFoundException e) {
+      return Response
+          .forStatus(
+              Status.NOT_FOUND.withReasonPhrase("Resource not found."));
+    } catch (IOException e) {
+      return Response
+          .forStatus(
+              Status.INTERNAL_SERVER_ERROR.withReasonPhrase("Failed to store resource."));
+    }
+
+    return Response.forStatus(Status.NO_CONTENT);
+  }
+
+  private Response<Resource> postResource(Request request) {
+    final Optional<ByteString> payload = request.payload();
+    if (!payload.isPresent()) {
+      return Response.forStatus(Status.BAD_REQUEST.withReasonPhrase("Missing payload."));
+    }
+
+    Resource resource;
+    try {
+      resource = OBJECT_MAPPER.readValue(payload.get().toByteArray(), Resource.class);
+    } catch (IOException e) {
+      return Response.forStatus(Status.BAD_REQUEST.withReasonPhrase("Invalid payload."));
+    }
```

Comment:
Use the EntityMiddleware to avoid having to do manual parsing and error handling.